### DISCUSSION TITLE
Running pipeline with various parameters

### DISCUSSION THREAD

**Discussion by TylerGrantSmith** [ropensci/targets#680] [Help]
Currently I have a very complex/long pipeline that depends on one target which is expected to be a single-valued parameter.  I would like to be able to execute this entire pipeline for a variety of parameters.  How do I do this with `targets`?  I have included an example pipeline below.   The third pipeline "works" but it doesn't seem ideal that it requires modifying all the targets to get it to work.

``` r
library(targets)
#> Warning: package 'targets' was built under R version 4.0.5

tar_dir({ 
  tar_script(
    list(
      tar_target(times, 1),
      tar_target(text, strrep('x', times)),
      tar_target(write_text, writeLines(text, paste0(times, ".txt")))
      ), ask = FALSE)
  tar_make()
  cat(readLines("1.txt"))
})
#> * start target times
#> * built target times
#> * start target text
#> * built target text
#> * start target write_text
#> * built target write_text
#> * end pipeline
#> Warning message:
#> package 'targets' was built under R version 4.0.5 
#> x

# doesn't work

tar_dir({ 
  tar_script(
    list(
      tar_target(times, c(1, 2)),
      tar_target(text, strrep('x', times)),
      tar_target(write_text, writeLines(text, paste0(times, ".txt")))
    ), ask = FALSE)
  tar_make()
  cat(readLines("1.txt")); cat("\n")
  cat(readLines("2.txt")); cat("\n")
})
#> * start target times
#> * built target times
#> * start target text
#> * built target text
#> * start target write_text
#> x error target write_text
#> * end pipeline
#> Error : invalid 'description' argument
#> In addition: Warning message:
#> package 'targets' was built under R version 4.0.5 
#> Error: callr subprocess failed: invalid 'description' argument
#> Visit https://books.ropensci.org/targets/debugging.html for debugging advice.

# works if I modify all the targets, but this is not ideal

tar_dir({ 
  tar_script(
    list(
      tar_target(times, c(1, 2)),
      tar_target(text, strrep('x', times), pattern = map(times)),
      tar_target(write_text, writeLines(text, paste0(times, ".txt")), pattern = map(text, times))
    ), ask = FALSE)
  tar_make()
  cat(readLines("1.txt")); cat("\n")
  cat(readLines("2.txt")); cat("\n")
})
#> * start target times
#> * built target times
#> * start branch text_819f0a1c
#> * built branch text_819f0a1c
#> * start branch text_0fc86902
#> * built branch text_0fc86902
#> * built pattern text
#> * start branch write_text_15b5fad6
#> * built branch write_text_15b5fad6
#> * start branch write_text_eb395f24
#> * built branch write_text_eb395f24
#> * built pattern write_text
#> * end pipeline
#> Warning message:
#> package 'targets' was built under R version 4.0.5 
#> x
#> xx
```

**Comment by wlandau** [comment]
Your last pipeline uses dynamic branching, which indeed does require modifying the pipeline. Static branching is another option, via functions `tarchetypes::tar_map()` and `tarchetypes::tar_eval()`: https://books.ropensci.org/targets/static.html. Static branching will change target names and thus invalidate the pipeline, but it will not require to modify the basic `tar_target()` statements.

**Comment by TylerGrantSmith** [comment]
@wlandau  Thanks for the feedback, and this does seem to get at what I wanted:

```
library(targets)
library(tarchetypes)

tar_dir({ 
  tar_script(
    tarchetypes::tar_map(
      list(times = c(1, 2)),
    list(
      # tar_target(times, 1),
      tar_target(text, strrep('x', times)),
      tar_target(write_text, writeLines(text, paste0(times, ".txt")))
    )), ask = FALSE)
  tar_make()
  cat(readLines("1.txt")); cat("\n")
  cat(readLines("2.txt")); cat("\n")
})
```

however, I still had to remove my `times` target so there wasn't a name conflict.  This is fine, I think, but it feels clumsy, as I have all my targets organized in separate files, and to effect this behavior would require me to define the `tar_map` in a different file from where the value actually gets used.  Is it not possible to add some kind of flag to the `tar_target(times, ...)` target so that `targets` recognizes the target should be used to develop static branches?  (I don't know if that terminology makes sense)

**Reply by wlandau** [reply]
I don't think I understand what you mean or what you are struggling with, but `targets` tries to do a lot so you do not have to micromanage files yourself. In the majority of cases, you can treat targets as R objects and let the package automatically store them to `_targets/objects/`.

**Reply by TylerGrantSmith** [reply]
Ya, I was afraid I couldn't explain the scenario well.  

Right now I have designed my pipeline by defining targets across multiple files and each file sources any other target files that it needs.  The goal was to have it be modular so that any any single file could be the script for a `tar_make` independently.  

I have created a mock-up in https://github.com/TylerGrantSmith/targets_mapping  The idea is that I can `tar_make(script = "A.r")` or `tar_make(script = "B.r")` .  Now, if I want to iterate `tar_make(script = "B.r")` using diferent values for `times` without modifying either `A.r` or `B.r`, then I have to do something like:

https://github.com/TylerGrantSmith/targets_mapping/blob/main/R/C1.r

I would like to do something like

https://github.com/TylerGrantSmith/targets_mapping/blob/main/R/C2.r

**Reply by wlandau** [reply]
So then is the issue trying to automate multiple versions of the same pipeline at the same time (branching and non-branching)? Seems like that would require manual modifications to the target script and techniques from https://books.ropensci.org/targets/projects.html, which you are already doing. Knowing what I know, I do not have a workaround to recommend, it is fine to comment/uncomment a target depending on what you need to do. Not sure I understand the scenarios this sort of friction would become time-consuming or prohibitive.

**Reply by TylerGrantSmith** [reply]
I see.  The friction is currently neither time-consuming nor prohibitive, but it is friction nonetheless.  Is what I was proposing in `C2.r` not possible?  i.e. is it not possible to modify `tar_map` to overload targets defined in `...` with values defined in `values`?

**Reply by wlandau** [reply]
So if you write something like this:

```{r}
# _targets.R
library(targets)
library(tarchetypes)
tar_map(
  list(times = c(1, 2)),
  list(
    tar_target(times, 1),
    tar_target(text, strrep('x', times)),
    tar_target(write_text, writeLines(text, paste0(times, ".txt"))
    )
  )
)
```

You will get an error:

```{r}
tar_visnetwork()
#> Error : target times cannot be in names(values).
```

The error message is important because the inclusion of `times` is a genuine bug in user-side code. If `tar_map()` were to automatically try to correct for this, the pipeline would be ambiguous, and frustrating/mysterious errors would pop up elsewhere.

Once you start using `tar_map()`, `times` only needs to be removed once, and then you're done. If you want to go back to a single value of `times`, it is easier to modify the `values` list to accomplish this rather than reinsert the `times` target.

**Reply by TylerGrantSmith** [reply]
It's a bug only in the sense that it currently doesn't allow that functionality.  If it were possible to explicitly say "I know what I am doing, override these values where they exist" then I don't see how that is ambiguous.  Part of the problem with commenting on `times` in my `A.r` is that now that pipeline is broken and it will not execute as an independent unit.

**Reply by wlandau** [reply]
Since you are already coding your pipeline modularly in different target lists, you could code it so the times target only appears in the non-tar_map() target lists. Just a different way of splitting up the target lists.

When a target conflicts with the values list, it is more often the case that the user just forgot a target was still there. This is often the case in large complex pipelines, and it is important to enforce guardrails early on.
